plugins/pipewire/PipeWireDevice.cpp

```diff
@@ -0,0 +1,300 @@
+		return;
+	}
+
+	spa_data& data = pw_buf->buffer->datas[0];
+	if(!data.data)
+	{
+		return;
+	}
+
+	spa_chunk* chunk = data.chunk;
+	if(!chunk)
+	{
+		return;
+	}
+
+	chunk->offset = 0;
+	chunk->stride = AUD_DEVICE_SAMPLE_SIZE(device->m_specs);
+	int n_frames = data.maxsize / chunk->stride;
+	if(pw_buf->requested)
+	{
+		n_frames = SPA_MIN(pw_buf->requested, n_frames);
+	}
+	chunk->size = n_frames * chunk->stride;
+
+	if(!device->m_playback)
+	{
+		memset(data.data, 0, AUD_FORMAT_SIZE(device->m_specs.format) * chunk->size);
+	}
+	else
+	{
+		device->mix((data_t*) data.data, n_frames);
+	}
+
+	AUD_pw_stream_queue_buffer(device->m_stream, pw_buf);
+}
+
+void PipeWireDevice::playing(bool playing)
+{
+	AUD_pw_thread_loop_lock(m_thread);
+	AUD_pw_stream_set_active(m_stream, playing);
+	AUD_pw_thread_loop_unlock(m_thread);
+
+	m_playback = playing;
+}
+
+PipeWireDevice::PipeWireDevice(const std::string& name, DeviceSpecs specs, int buffersize) : m_synchronizer(this), m_playback(false)
+{
+	if(specs.channels == CHANNELS_INVALID)
+		specs.channels = CHANNELS_STEREO;
+	if(specs.format == FORMAT_INVALID)
+		specs.format = FORMAT_FLOAT32;
+	if(specs.rate == RATE_INVALID)
+		specs.rate = RATE_48000;
+
+	m_specs = specs;
+	spa_audio_format format = SPA_AUDIO_FORMAT_F32;
+	switch(m_specs.format)
+	{
+	case FORMAT_U8:
+		format = SPA_AUDIO_FORMAT_U8;
+		break;
+	case FORMAT_S16:
+		format = SPA_AUDIO_FORMAT_S16;
+		break;
+	case FORMAT_S24:
+		format = SPA_AUDIO_FORMAT_S24;
+		break;
+	case FORMAT_S32:
+		format = SPA_AUDIO_FORMAT_S32;
+		break;
+	case FORMAT_FLOAT32:
+		format = SPA_AUDIO_FORMAT_F32;
+		break;
+	case FORMAT_FLOAT64:
+		format = SPA_AUDIO_FORMAT_F64;
+		break;
+	default:
+		break;
+	}
+
+	AUD_pw_init(nullptr, nullptr);
+
+	m_thread = AUD_pw_thread_loop_new(name.c_str(), nullptr);
+	if(!m_thread)
+	{
+		AUD_THROW(DeviceException, "Could not create PipeWire thread.");
+	}
+
+	m_events = std::make_unique<pw_stream_events>();
+	m_events->version = PW_VERSION_STREAM_EVENTS;
+	m_events->state_changed = PipeWireDevice::handle_state_changed;
+	m_events->process = PipeWireDevice::mix_audio_buffer;
+
+	pw_properties *stream_props = AUD_pw_properties_new(
+				PW_KEY_MEDIA_TYPE, "Audio",
+				PW_KEY_MEDIA_CATEGORY, "Playback",
+				PW_KEY_MEDIA_ROLE, "Production",
+				NULL);
+
+	/* Set the requested sample rate and latency. */
+	AUD_pw_properties_setf(stream_props, PW_KEY_NODE_RATE, "1/%u", uint(m_specs.rate));
+	AUD_pw_properties_setf(stream_props, PW_KEY_NODE_LATENCY, "%u/%u", buffersize, uint(m_specs.rate));
+
+	m_stream = AUD_pw_stream_new_simple(
+			AUD_pw_thread_loop_get_loop(m_thread),
+			name.c_str(),
+			stream_props,
+			m_events.get(),
+			this);
+	if(!m_stream)
+	{
+		AUD_pw_thread_loop_destroy(m_thread);
+		AUD_THROW(DeviceException, "Could not create PipeWire stream.");
+	}
+
+	spa_audio_info_raw info{};
+	info.channels = m_specs.channels;
+	info.format = format;
+	info.rate = m_specs.rate;
+
+	const spa_pod *params[1];
```

Comment:
What's the reasoning behind having an array here? A single pointer should suffice (using `&params` below instead of params).

Reply:
I simply copy pasted some of the Pipewire code examples for this part, so no reasoning really.
Will change this to not be an array as you suggested.